### PR TITLE
Solve pydantic deprecation warnings

### LINKED ISSUE #12654: [BUG] Pydantic validator and root_validator are deprecated
### Issues Policy acknowledgement

- [X] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### Willingness to contribute

Yes. I would be willing to contribute a fix for this bug with guidance from the MLflow community.

### MLflow version

- Client: 2.14.2

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: macOS Sonoma 14.5
- **Python version**: 3.11.9
- **yarn version, if running the dev UI**: - 

### Describe the problem

Pydantic's `validator` and `root_validator` are deprecated and should be changed to `field_validator` and `model_validator`, respectively.

### Tracking information

```shell
REPLACE_ME
```

### Code to reproduce issue

```
import mlflow
```

### Stack trace

```
/opt/homebrew/Caskroom/mambaforge/base/envs/<virual_environemnt>/lib/python3.11/site-packages/mlflow/gateway/config.py:412: PydanticDeprecatedSince20: Pydantic V1 style `@validator` validators are deprecated. You should migrate to Pydantic V2 style `@field_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.8/migration/
    @validator("route_type", pre=True)

/opt/homebrew/Caskroom/mambaforge/base/envs/<virtual_environment>/lib/python3.11/site-packages/mlflow/gateway/config.py:390: PydanticDeprecatedSince20: Pydantic V1 style `@root_validator` validators are deprecated. You should migrate to Pydantic V2 style `@model_validator` validators, see the migration guide for more details. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.8/migration/
    @root_validator(skip_on_failure=True)
```

### Other info / logs

```
REPLACE_ME
```

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by MariusMerkleQC**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/MariusMerkleQC/mlflow/pull/12698?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/12698/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 12698
```

</p>
</details>

### Related Issues/PRs

Resolve #12654 

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [ ] No (this PR will be included in the next minor release)

### CODE REVIEW

File: dev/set_matrix.py [hunk 1]
```diff
@@ -269,11 +307,15 @@ def get_requires_python(package: str, version: str) -> str:
     return next((c for c in candidates if spec.contains(c)), None) or candidates[0]
 
 
-def get_python_version(python: Optional[Dict[str, str]], package: str, version: str) -> str:
+def get_python_version(
+    python: Optional[Dict[str, str]], package: str, version: str
+) -> str:
```
Comment (B-Step62) [inline dev/set_matrix.py:312]: Can you revert these unrelated formatting?

**Comment by B-Step62** [comment]
@MariusMerkleQC The change looks good to me, but there are a few formatting issues. Could you run our pre-commit hook to ensure the correct format, and remove formatting in unrelated files?

Here is the guide for installing and running the pre-commit hook:
* Installation: https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.md#manual-python-development-environment-configuration
* Running: https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.md#running-python-tests

**Comment by MariusMerkleQC** [comment]
I'd like to, but I have trouble following the [installation guide](https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.md#manual-python-development-environment-configuration): when running the command `pre-commit install -t pre-commit -t prepare-commit-msg`, I get the error `zsh: command not found: pre-commit`.

**Comment by B-Step62** [comment]
@MariusMerkleQC Could you try `pip install pre-commit`?

**Comment by MariusMerkleQC** [comment]
Yes, and then I can run `pre-commit install -t pre-commit -t prepare-commit-msg`. And when running `pre-commit run --all-files`, the output looks as follows:

image

**Comment by B-Step62** [comment]
@MariusMerkleQC You don't have the linter packages installed. Please run `pip install -r requirements/lint-requirements.txt`.
Btw these environment setup is described in [the contribution guide](https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.md#developing-and-testing-mlflow), would you mind following this first?

**Comment by MariusMerkleQC** [comment]
Sorry for the additional overhead, but when following the [setup guide](https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.md#manual-python-development-environment-configuration) step-by-step, I ran into these issues:

- `pip install -r requirements/lint-requirements.txt` will not work if `pip install pre-commit` is not run before
- I cannot find the command `pip install -r requirements/lint-requirements.txt` in the setup guide.

**Comment by B-Step62** [comment]
@MariusMerkleQC I would recommend using [the automated set up script](https://github.com/mlflow/mlflow/blob/master/CONTRIBUTING.md#automated-python-development-environment-configuration) to create the complete virtual environment.

Alternatively, you can attach `autoformat` label to the PR to run formatting script via Github action.

File: mlflow/gateway/config.py [hunk 2]
```diff
@@ -366,71 +477,175 @@ class RouteConfig(AliasedConfigModel):
...
+                        "A config must be supplied when setting a provider. The provider entry for "
+                        f"{model_instance.provider} is incorrect."
+                    )
+            return model
+
+    if IS_PYDANTIC_V2:
+        from pydantic import model_validator as _model_validator
+
+        @_model_validator(mode="after")
+        def validate_route_type_and_model_name(self) -> Self:
```
Comment (B-Step62) [inline mlflow/gateway/config.py:540]:
```suggestion
        def validate_route_type_and_model_name(self):
```


The doc building job is also failing due to `ImportError: cannot import name 'Self' from 'typing' (/home/circleci/.pyenv/versions/3.8.16/lib/python3.8/typing.py)`. It seems the `Self` type is not supported by the our current version of typing extension.

File: mlflow/gateway/config.py [hunk 3]
```diff
@@ -72,26 +83,59 @@ class RouteType(str, Enum):
...
+        @classmethod
+        def validate_cohere_api_key(cls, value):
+            return _resolve_api_key_from_input(value)
+
+    else:
+        from pydantic import validator as _validator
+
+        @_validator("cohere_api_key", pre=True)
+        def validate_cohere_api_key(cls, value):
+            return _resolve_api_key_from_input(value)
```
Comment (B-Step62) [inline mlflow/gateway/config.py:99]: What about defining a new decorator to avoid repeating this if-else logic all over the place?

For example,
```
def pydantic_field_validator(field_name: str):
    def decorator(func: Callable) -> Callable:
        if IS_PYDANTIC_V2:
            from pydantic import field_validator
            return field_validator(field_name, mode="before")(func)
        else:
            from pydantic import validator
            return validator(field_name, pre=True)(func)

    return decorator 
```

then we can do

```
    @pydantic_field_validator("cohere_api_key")
    def validate_cohere_api_key(cls, value):
        return _resolve_api_key_from_input(value)
```

**Comment by c0dearm** [comment]
Since this PR was stuck because of the linting issues, I've opened a new one which has the same changes, but rebased and with the linting fixed: https://github.com/mlflow/mlflow/pull/13023
